### ISSUE TITLE
Boot and the data_readers.clj file

### ISSUE THREAD

**Issue by Malabarba** [clojure-emacs/cider#1462]
@Naomarik reports [here](https://github.com/clojure-emacs/cider/issues/1218#issuecomment-163563511) that the debugger isn't working in boot. The error indicates that boot doesn't see our `data_readers.clj` file at the root of the class path.

According to http://clojure.org/reader:

>When Clojure starts, it searches for files named data_readers.clj at the root of the classpath. Each such file must contain a Clojure map of symbols, like this:
>
>    {foo/bar my.project.foo/bar
>     foo/baz my.project/baz}

Does anyone know if this might be a boot bug or if we're doing something wrong?

**Comment by bbatsov** [comment]
Pinging @micha.

**Comment by Naomarik** [comment]
I've found that `(boot.core/load-data-readers!)` will work once the repl is going, but I'm not sure where to put this so it does it automatically.
Reference: https://github.com/boot-clj/boot/issues/47

**Comment by Malabarba** [comment]
Maybe cider-nrepl should do it? Though I'm not sure how.

**Comment by Malabarba** [comment]
Although, considering that this is a general Clojure thing, it would be nice if Boot handled it by default.

**Comment by camelpunch** [comment]
Bump. I've found that `(boot.core/load-data-readers!)` removes the error, but doesn't cause C-M-x to jump into the debugger when a `#break` is present.

**Comment by Malabarba** [comment]
@camelpunch There are some places where a `#break` doesn't work, so that might be unrelated to boot. Could you see if it works for you by hitting `C-M-x` in the following snippet?

```clj
(let [x 1]
  (inc #break x))
```

**Comment by camelpunch** [comment]
Yes, that works. Thanks!

**Comment by Malabarba** [comment]
If anyone here uses boot, I'd still like to know if there's any way to load forms on startup. Even if there's no way for cider to do it, we should add instructions to the Readme.

@alexander-yakushev You've been using boot lately, haven't you?

**Comment by alexander-yakushev** [comment]
I can confirm that doing `(boot.core/load-data-readers)` fixes the problem and the debugger works. Boot only provides an option to provide `:eval` form but only for the client — it doesn't work if REPL is started in server mode.

@Malabarba RE: debugger. Holy maker of pepperoni, I just tried it for the first time and it's frigging amazing! I didn't know I've been missing so much :).

**Comment by Malabarba** [comment]
@alexander-yakushev Thanks ^^

I guess the best we can do then is load this form as part of cider-nrepl if the symbol `boot.core/load-data-readers` resolves to a var. @bbatsov what do you think?

**Comment by bbatsov** [comment]
I think it's a reasonable approach and I've heard no other suggestions, so let's get this done.

**Comment by alexander-yakushev** [comment]
I object, your honor.

First, boot.core/load-data-readers does nothing more but reloading readers
and setting root of *data-readers* to it.

Second, that fails when evaluated at the same time as middleware init,
because apparently Boot runs those in a separate thread, and I get 'setting
from non-binder thread' exception.

**Comment by bbatsov** [comment]
OK, I guess we're back to the drawing board then. :-)

**Comment by bbatsov** [comment]
@Malabarba Maybe we should at least document this problem somewhere?

**Comment by micha** [comment]
If there's anything I can do to help let me know. (I'm mostly a Vim user
though.)

--
Micha Niskin

**Comment by alexander-yakushev** [comment]
@micha How about having a separate `:eval` option that is evaluated by server rather than the one invoked by REPL client?

**Comment by Malabarba** [comment]
@micha It would be great if `data_readers.clj` was loaded by default when a boot REPL is started. The Clojure documentation states that this file is loaded automatically, so I think it makes sense for the REPL to behave the same way.

**Comment by bbatsov** [comment]
Yeah, that'd be a dream solution. :-)

**Comment by micha** [comment]
Sure, I can do that. Would you consider it a bugfix or feature? So I know
whether to land it in 2.6.0 or a patch release.

--
Micha Niskin

**Comment by Malabarba** [comment]
I think I'd call it a bugfix, but it could definitely  go either way.

**Comment by bbatsov** [comment]
Any progress with this?

**Comment by Malabarba** [comment]
While we wait for an upstream fix, I think I have a workaround. the debug.clj middleware can call `(boot.core/load-data-readers!)` when it receives the `debug-init` message. That only happens after the repl is connected, so I think it avoids the previous problem, right?

(of course, it will check whether that function exists first, so it won't error on non-boot projects)

**Comment by alexander-yakushev** [comment]
> (of course, it will check whether that function exists first, so it won't error on non-boot projects)

There isn't much point in calling exactly Boot's function as it is nothing else but:

```clojure
  (#'clojure.core/load-data-readers)
  (set! *data-readers* (.getRawRoot #'*data-readers*))
```

And calling that in Leiningen won't hurt either.

**Comment by Malabarba** [comment]
Even better.

**Comment by Malabarba** [comment]
Pushed Alex's snippet.

**Comment by tmarble** [comment]
Clojars is still based on the version https://github.com/clojure-emacs/cider-nrepl/commit/1723057b7dcba57fc10108b0b38f0e6a5293df6d

I just did a local `lein install` with https://github.com/clojure-emacs/cider-nrepl/commit/ebf352681fec7f378cf078fae5d62f4c1502798c  and that avoids the ugly stack trace (`Can't set!: *data-readers* from non-binding thread`)

FWIW I don't think the debugger is actually working as I tried `C-u C-M-x`, but that provoked a stack trace.

**Comment by Malabarba** [comment]
Sorry, I've redeployed now. If the debugger still doesn't work for you, please file a new issue.

**Comment by Malabarba** [comment]
Ignore the accidental close, the issue itself is very much still open.

**Comment by Malabarba** [comment]
@micha Has this made into a boot stable release?

**Comment by arichiardi** [comment]
@Malabarba It's been [milestoned](https://github.com/boot-clj/boot/milestones/2.6.0)

**Comment by Malabarba** [comment]
@arichiardi Thanks for the notice!

**Comment by micha** [comment]
Attempted a fix with boot-clj/boot#439...does it work?

**Comment by bbatsov** [comment]
Guess someone should try this out.

**Comment by tmarble** [comment]
@micha OMG that works!

@Malabarba is there a current "cheatsheet" for the debugger? I've seen your blog, but I'm wondering if there is a current list keybindings (functions) and overview of what's possible? Sorry if this is obvious and I've missed it (ENOTENOUGHCOFFEE).

@bbatsov I just have to say CIDER is *amazing*! I love the docstring on mouseover symbol thing!

**Comment by bbatsov** [comment]
> @Malabarba is there a current "cheatsheet" for the debugger? I've seen your blog, but I'm wondering if there is a current list keybindings (functions) and overview of what's possible? Sorry if this is obvious and I've missed it (ENOTENOUGHCOFFEE).

See http://cider.readthedocs.org/en/latest/extended_workflow/#debugging :-)

**Comment by tmarble** [comment]
@bbatsov Благодаря!!!

**Comment by Malabarba** [comment]
@tmarble indeed. The man page should always contain everything. And we try to keep the debugger features easily discoverable.

**Comment by Malabarba** [comment]
Ok. Since the fixed has been merged into boot master, I'm closing this issue. We should have a definitive fix when 2.6.0 comes out.

**Comment by zalky** [comment]
Hi all, is anyone still experiencing this issue?

I can load all the cider data readers:

```clj
boot.user> (load-data-readers!)
{db/id #'datomic.db/id-literal,
 db/fn #'datomic.function/construct,
 base64 #'datomic.codec/base-64-literal,
 ordered/set #'flatland.ordered.set/into-ordered-set,
 ordered/map #'flatland.ordered.map/ordered-map,
 dbg #'cider.nrepl.middleware.debug/debug-reader,     ; <- appears to be there
 break #'cider.nrepl.middleware.debug/breakpoint-reader,
 light #'cider.nrepl.middleware.enlighten/light-reader}
```
and I still get an error when the file is compiled:

```
clojure.lang.Compiler$CompilerException: java.lang.RuntimeException: No reader function for tag dbg
```
Cider version: 0.18.0
Boot version: 2.8.1
